text/3861-iter-macro.md

```diff
@@ -0,0 +1,496 @@
+The subsequent call to `iter.next()` will resume executing the iterator immediately after that `yield` expression.
+
+When the iterator exits the closure's body, such as by executing to the end of the body or by executing a `return` expression, the corresponding call to `next()` returns `None`, indicating the end of the iterator.
+
+The body of an iterator closure must have a return value compatible with the `()`.
+In other words, it must either return `()` or never return.
+
+Auto traits such as `Send` and `Sync` are computed separately for the iterator closure and the iterator returned by the closure.
+For iterator closures, these are computed the same as for any other closure.
+For the corresponding iterator, auto traits are computed based on the values that must be saved across a `yield` point, similar to how auto traits for an `async` block are computed based on the values that are saved across `await` points.
+
+It is not possible to hold a borrow of a value local to the iterator across a `yield` point.
+Similarly, it is not possible to `yield` a reference to an iterator-local value.
+Note that holding and yielding references that completely outlive the iterator closure (such as a reference captured from the environment) are allowed.
+
+# Drawbacks
+
+- Iterator closures have some limitations around self-borrows and lending.
+- The macro-based syntax looks less built-in than if it were actually built-in.
+- Does not exactly support the plain iterator block use case. You always have to create and invoke a closure.
+
+# Rationale and alternatives
+
+This incremental step forward will help us make continued progress towards more advanced generators.
+If we do not take this step, we expect we will see little further development in generators and users will remain frustrated that they do not have convenient syntax for creating iterators.
+This is a feature that is often missed by Rust programmers.
+
+## Why Iterator Closures?
+
+It might seem more obvious to have `iter!` evaluate directly to an `impl Iterator` with no intermediate closure step.
+We instead recommend returning an iterator closure.
+This is largely as a result of what we have learned from our experience with `async`.
+
+Having a two step process between creating the iterator-like object and beginning iteration allows us to support scenarios such as where the result of `iter!` is `Send` but the iterator is no longer `Send` once iteration starts.
+See Yosh Wuyts' [The Gen Auto-Trait Problem] for more details.
+In `async`, we've had a lot of discussion about using `IntoFuture` for this two stage process but decided that it is better represented through async closures.
+For iterators and generators, we'd like to set the same precedent from the beginning.
+
+[The Gen Auto-Trait Problem]: https://blog.yoshuawuyts.com/gen-auto-trait-problem/
+
+Second, having convenient syntax for creating inline iterators will create an incentive to create more powerful combinators.
```

Comment:
Can you elaborate more on what kinds of things these "more powerful combinators" would be?  Where does the `Fn`-ness come in?

(I understand the convenience of things like <https://doc.rust-lang.org/stable/std/iter/fn.zip.html>, but don't know enough to bootstrap that to "more powerful".)